File: src/ContosoAirlinePOS/ContosoAirlinePOSCs/ContosoAirlinePOSCs/ContosoAirlinePOSCs/Models/Flight.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ContosoAirlinePOSCs.Models
{
    public class Flight : ModelBase
    {
        private int _flightId;
        public int FlightId
        {
            get { return _flightId; }
            set { _flightId = value; OnPropertyChanged(); }
        }

        private string _flightCode;
        public string FlightCode
        {
            get { return _flightCode; }
            set { _flightCode = value; OnPropertyChanged(); }
        }


        private Airport _source;
        public Airport Source
        {
            get { return _source; }
            set { _source = value; OnPropertyChanged(); }
        }

        private Airport _destination;
        public Airport Destination
        {
            get { return _destination; }
            set { _destination = value; OnPropertyChanged(); }
        }

        private DateTime _departure;
        public DateTime Departure
        {
            get { return _departure; }
            set { _departure = value; OnPropertyChanged(); }
        }

        private DateTime _arrival;
        public DateTime Arrival
        {
            get { return _arrival; }
            set { _arrival = value; OnPropertyChanged(); }
        }

        private TimeSpan _duration;
        public TimeSpan Duration
        {
            get { return _duration; }
            set { _duration = value; OnPropertyChanged(); }
        }

        private int _hops;
        public int Hops
        {
            get { return _hops; }
            set { _hops = value; OnPropertyChanged(); }
        }

        private List<Booking> _bookings;
        public List<Booking> Bookings
        {
            get { return _bookings; }
            set { _bookings = value; OnPropertyChanged(); }
        }

        private double _ticketPrice;
        public double TicketPrice
        {
            get { return _ticketPrice; }
            set { _ticketPrice = value; OnPropertyChanged(); }
        }

        private string _flightStatus;
        public string FlightStatus
        {
            get { return _flightStatus; }
            set { _flightStatus = value; OnPropertyChanged(); }
        }

        private string _gate;
        public string Gate
        {
            get { return _gate; }
            set { _gate = value; OnPropertyChanged(); }
        }


        public Flight Copy()
        {
            Flight flight = new Flight()
            {
                FlightCode = this.FlightCode,
                Departure = this.Departure,
                Arrival = this.Arrival,
                Source = this.Source,
                Destination = this.Destination,
                Duration = this.Duration,
                Hops = this.Hops
            };

            return flight;
        }
    }
}

File: src/ContosoAirlinePOS/ContosoAirlinePOSCpp/ContosoAirlinePOSCpp/ContosoAirlinePOSCpp/Flight.cpp
#include "pch.h"
#include "Flight.h"
#include "Flight.g.cpp"

namespace winrt::ContosoAirlinePOSCpp::implementation
{
    hstring Flight::FlightId()
    {
        return m_flightID;
    }
    void Flight::FlightId(hstring const& value)
    {
        if (m_flightID != value)
        {
            m_flightID = value;
            m_propertyChanged(*this, Microsoft::UI::Xaml::Data::PropertyChangedEventArgs{ L"FlightId" });
        }
    }
    hstring Flight::FlightCode()
    {
        return m_flightCode;
    }
    void Flight::FlightCode(hstring const& value)
    {
        if (m_flightCode != value)
        {
            m_flightCode = value;
            m_propertyChanged(*this, Microsoft::UI::Xaml::Data::PropertyChangedEventArgs{ L"FlightCode" });
        }
    }
    hstring Flight::Departure()
    {
        return m_departure;
    }
    void Flight::Departure(hstring const& value)
    {
        if (m_departure != value)
        {
            m_departure = value;
            m_propertyChanged(*this, Microsoft::UI::Xaml::Data::PropertyChangedEventArgs{ L"Departure" });
            m_propertyChanged(*this, Microsoft::UI::Xaml::Data::PropertyChangedEventArgs{ L"DepartureDate" });
            m_propertyChanged(*this, Microsoft::UI::Xaml::Data::PropertyChangedEventArgs{ L"DepartureTime" });
        }
    }
    hstring Flight::Arrival()
    {
        return m_arrival;
    }
    void Flight::Arrival(hstring const& value)
    {
        if (m_arrival != value)
        {
            m_arrival = value;
            m_propertyChanged(*this, Microsoft::UI::Xaml::Data::PropertyChangedEventArgs{ L"Arrival" });
            m_propertyChanged(*this, Microsoft::UI::Xaml::Data::PropertyChangedEventArgs{ L"ArrivalDate" });
            m_propertyChanged(*this, Microsoft::UI::Xaml::Data::PropertyChangedEventArgs{ L"ArrivalTime" });
        }
    }
    hstring Flight::Duration()
    {
        return m_duration;
    }
    void Flight::Duration(hstring const& value)
    {
        if (m_duration != value)
        {
            m_duration = value;
            m_propertyChanged(*this, Microsoft::UI::Xaml::Data::PropertyChangedEventArgs{ L"Duration" });
        }
    }
    int32_t Flight::Hops()
    {
        return m_hops;
    }
    void Flight::Hops(int32_t value)
    {
        if (m_hops != value)
        {
            m_hops = value;
            m_propertyChanged(*this, Microsoft::UI::Xaml::Data::PropertyChangedEventArgs{ L"Hops" });
        }
    }
    ContosoAirlinePOSCpp::Airport Flight::Source()
    {
        return m_source;
    }
    void Flight::Source(ContosoAirlinePOSCpp::Airport const& value)
    {
        if (m_source != value)
        {
            m_source = value;
            m_propertyChanged(*this, Microsoft::UI::Xaml::Data::PropertyChangedEventArgs{ L"Source" });
        }
    }
    ContosoAirlinePOSCpp::Airport Flight::Destination()
    {
        return m_destination;
    }
    void Flight::Destination(ContosoAirlinePOSCpp::Airport const& value)
    {
        if (m_destination != value)
        {
            m_destination = value;
            m_propertyChanged(*this, Microsoft::UI::Xaml::Data::PropertyChangedEventArgs{ L"Destination" });
        }
    }
    double Flight::TicketPrice()
    {
        return m_ticketPrice;
    }
    void Flight::TicketPrice(double value)
    {
        if (m_ticketPrice != value)
        {
            m_ticketPrice = value;
            m_propertyChanged(*this, Microsoft::UI::Xaml::Data::PropertyChangedEventArgs{ L"TicketPrice" });
        }
    }
    winrt::event_token Flight::PropertyChanged(Microsoft::UI::Xaml::Data::PropertyChangedEventHandler const& handler)
    {
        return m_propertyChanged.add(handler);
    }
    void Flight::PropertyChanged(winrt::event_token const& token) noexcept
    {
        m_propertyChanged.remove(token);
    }
    hstring Flight::FlightStatus()
    {
        return m_flightStatus;
    }
    void Flight::FlightStatus(hstring const& value)
    {
        if (m_flightStatus != value)
        {
            m_flightStatus = value;
            m_propertyChanged(*this, Microsoft::UI::Xaml::Data::PropertyChangedEventArgs{ L"FlightStatus" });
        }
    }
    hstring Flight::Gate()
    {
        return m_gate;
    }
    void Flight::Gate(hstring const& value)
    {
        if (m_gate != value)
        {
            m_gate = value;
            m_propertyChanged(*this, Microsoft::UI::Xaml::Data::PropertyChangedEventArgs{ L"Gate" });
        }
    }
    hstring Flight::DepartureDate()
    {
        std::string strTime;
        std::string strDate;
        SplitDateTime(to_string(m_departure), strTime, strDate);

        return to_hstring(strDate);
    }
    hstring Flight::ArrivalDate()
    {
        std::string strTime;
        std::string strDate;
        SplitDateTime(to_string(m_arrival), strTime, strDate);

        return to_hstring(strDate);
    }
    hstring Flight::DepartureTime()
    {
        std::string strTime;
        std::string strDate;
        SplitDateTime(to_string(m_departure), strTime, strDate);

        return to_hstring(strTime);
    }
    hstring Flight::ArrivalTime()
    {
        std::string strTime;
        std::string strDate;
        SplitDateTime(to_string(m_arrival), strTime, strDate);

        return to_hstring(strTime);
    }
    void Flight::SplitDateTime(std::string strInput, std::string& strTime, std::string& strDate)
    {
        //"2020-08-19.06:00:00"
        if (!strInput.empty())
        {
            int nIndex = (int)strInput.rfind(".");
            if (nIndex != 0)
            {
                std::string strNewString = strInput.substr(nIndex + 1);
                if (!strNewString.empty())
                {
                    strTime = strNewString;
                }               
                std::string strNewString1 = strInput.substr(0 , nIndex);
                if (!strNewString1.empty())
                {
                    strDate = strNewString1;
                }
            }
        }
    }
}
